src/include/duckdb/storage/compression/patas/patas_analyze.hpp

```diff
@@ -0,0 +1,138 @@
+		if (!HasEnoughSpace()) {
+			StartNewSegment();
+		}
+		patas::PatasCompression<EXACT_TYPE, true>::Store(value, state.patas_state);
+		previous_value = value;
+		group_idx++;
+		if (group_idx == PatasPrimitives::PATAS_GROUP_SIZE) {
+			StartNewGroup();
+		}
+	}
+
+	idx_t CurrentGroupMetadataSize() const {
+		idx_t metadata_size = 0;
+
+		// Offset to the data of the group
+		metadata_size += sizeof(uint32_t);
+		// Packed Trailing zeros + significant bytes + index_offsets for group
+		metadata_size += 2 * group_idx;
+		return metadata_size;
+	}
+
+	void StartNewSegment() {
+		StartNewGroup();
+		data_byte_size += UsedSpace();
+		metadata_byte_size += PatasPrimitives::HEADER_SIZE;
+		state.patas_state.byte_writer.SetStream(nullptr);
+	}
+
+	idx_t RequiredSpace() const {
+		idx_t required_space = 0;
+		required_space += sizeof(EXACT_TYPE);
+		required_space += sizeof(uint16_t);
+		return required_space;
+	}
+
+	void StartNewGroup() {
+		previous_value = 0;
+		metadata_byte_size += CurrentGroupMetadataSize();
+		group_idx = 0;
+		state.patas_state.Reset();
+	}
+
+	idx_t UsedSpace() const {
+		return state.patas_state.byte_writer.BytesWritten();
+	}
+
+	bool HasEnoughSpace() {
+		idx_t total_bytes_used = 0;
+		total_bytes_used += AlignValue(PatasPrimitives::HEADER_SIZE + UsedSpace() + RequiredSpace());
+		total_bytes_used += CurrentGroupMetadataSize();
+		total_bytes_used += metadata_byte_size;
+		return total_bytes_used <= Storage::BLOCK_SIZE;
+	}
+
+	idx_t TotalUsedBytes() const {
+		return metadata_byte_size + AlignValue(data_byte_size + UsedSpace());
+	}
+};
+
+struct EmptyPatasWriter {
+
+	template <class VALUE_TYPE>
+	static void Operation(VALUE_TYPE uncompressed_value, bool is_valid, void *state_p) {
+		using EXACT_TYPE = typename FloatingToExact<VALUE_TYPE>::type;
+
+		auto state_wrapper = (PatasAnalyzeState<VALUE_TYPE> *)state_p;
+		state_wrapper->WriteValue(*(EXACT_TYPE *)(&uncompressed_value), is_valid);
+	}
+};
+
+template <class T>
+unique_ptr<AnalyzeState> PatasInitAnalyze(ColumnData &col_data, PhysicalType type) {
+	return make_unique<PatasAnalyzeState<T>>();
+}
+
+template <class T>
+bool PatasAnalyze(AnalyzeState &state, Vector &input, idx_t count) {
+	auto &analyze_state = (PatasAnalyzeState<T> &)state;
+	UnifiedVectorFormat vdata;
+	input.ToUnifiedFormat(count, vdata);
+
+	auto data = (T *)vdata.data;
+	for (idx_t i = 0; i < count; i++) {
+		auto idx = vdata.sel->get_index(i);
+		analyze_state.state.template Update<EmptyPatasWriter>(data[idx], vdata.validity.RowIsValid(idx));
+	}
+	return true;
+}
+
+template <class T>
+idx_t PatasFinalAnalyze(AnalyzeState &state) {
+	auto &patas_state = (PatasAnalyzeState<T> &)state;
+	// Finish the last "segment"
+	patas_state.StartNewSegment();
+	const auto final_analyze_size = patas_state.TotalUsedBytes();
+	// printf("ANALYZE: ROWGROUP_SIZE: %llu\n", final_analyze_size);
+	return final_analyze_size;
```

Comment:
Both Patas and Chimp return the total used bytes, which means Patas will almost never be chosen, as it generally has a slightly worse compression ratio. Could we penalize Chimp (and perhaps penalize Patas slightly over uncompressed)? e.g. Chimp could have a `*2` multiplier, and Patas a `*1.2` or so.

Reply:
Yes, I remember you mentioned I should do this, just didn't really click how yet.

But this way makes perfect sense, I'll add it :thumbsup:

Reply:
I think we actually want to have this function return a struct at some point in the future, e.g. something like:

```sql
struct CompressionInfo {
   idx_t compression_speed;
   idx_t decompression_speed;
   idx_t estimated_size;
};
```

That way, we can check what we care about ourselves and add e.g. a configuration. We could also do stuff like, "for temporary tables we care less about estimated size and more about decompression speed than for persistent tables", etc.

Anyway, future PR :)